File: storage/validator.go
package storage

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"image"
	_ "image/jpeg"
	_ "image/png"
	"io"
	"net/http"
	"path/filepath"
	"strings"
)

type FileValidationResult struct {
	OriginalFilename string
	MIMEType         string
	Extension        string
	FileSize         int64
	FileHash         string
	Width            *int
	Height           *int
}

type FileValidator struct {
	MaxSizeBytes      int64
	MinImageDimension int
}

func NewFileValidator(maxSizeBytes int64, minImageDimension int) *FileValidator {
	if maxSizeBytes <= 0 {
		maxSizeBytes = 10 * 1024 * 1024 // 10MB default
	}
	if minImageDimension <= 0 {
		minImageDimension = 300
	}
	return &FileValidator{
		MaxSizeBytes:      maxSizeBytes,
		MinImageDimension: minImageDimension,
	}
}

func (v *FileValidator) Validate(file io.ReadSeeker, originalFilename string, declaredSize int64) (*FileValidationResult, error) {
	// 1. Check extension
	ext := strings.ToLower(filepath.Ext(originalFilename))
	blockedExtensions := map[string]bool{
		".exe": true, ".sh": true, ".bat": true, ".cmd": true,
		".js": true, ".py": true, ".php": true, ".zip": true,
		".tar": true, ".gz": true, ".rar": true, ".7z": true,
	}
	if blockedExtensions[ext] {
		return nil, fmt.Errorf("file type '%s' is strictly prohibited for security reasons", ext)
	}

	allowedExtensions := map[string]bool{
		".pdf": true, ".jpg": true, ".jpeg": true, ".png": true,
	}
	if !allowedExtensions[ext] {
		return nil, fmt.Errorf("unsupported file extension '%s'. Only PDF, JPG, JPEG, and PNG files are allowed", ext)
	}

	// 2. Check Size
	if declaredSize > v.MaxSizeBytes {
		return nil, fmt.Errorf("file size exceeds maximum limit of %d MB", v.MaxSizeBytes/(1024*1024))
	}

	// 3. Read header to detect actual MIME type
	buffer := make([]byte, 512)
	n, err := file.Read(buffer)
	if err != nil && !errors.Is(err, io.EOF) {
		return nil, fmt.Errorf("failed to read file header: %w", err)
	}

	detectedMIME := http.DetectContentType(buffer[:n])
	// Adjust for PDF or custom mime checks
	if ext == ".pdf" && !strings.Contains(detectedMIME, "pdf") && !strings.Contains(detectedMIME, "application/octet-stream") {
		// Verify PDF magic bytes '%PDF'
		if n >= 4 && string(buffer[:4]) == "%PDF" {
			detectedMIME = "application/pdf"
		} else {
			return nil, errors.New("file content does not match valid PDF format")
		}
	}

	allowedMIMEs := map[string]bool{
		"application/pdf": true,
		"image/jpeg":      true,
		"image/jpg":       true,
		"image/png":       true,
	}

	if !allowedMIMEs[detectedMIME] && !strings.HasPrefix(detectedMIME, "image/") {
		return nil, fmt.Errorf("invalid file MIME type '%s'", detectedMIME)
	}

	// 4. Compute SHA-256 Hash and Size
	if _, err := file.Seek(0, io.SeekStart); err != nil {
		return nil, err
	}

	hasher := sha256.New()
	size, err := io.Copy(hasher, file)
	if err != nil {
		return nil, fmt.Errorf("failed to compute file hash: %w", err)
	}

	if size > v.MaxSizeBytes {
		return nil, fmt.Errorf("file size %d bytes exceeds maximum limit of %d MB", size, v.MaxSizeBytes/(1024*1024))
	}

	fileHash := hex.EncodeToString(hasher.Sum(nil))

	// 5. Image Resolution Check (if image)
	var widthPtr, heightPtr *int
	if strings.HasPrefix(detectedMIME, "image/") {
		if _, err := file.Seek(0, io.SeekStart); err == nil {
			config, _, err := image.DecodeConfig(file)
			if err == nil {
				w, h := config.Width, config.Height
				if w < v.MinImageDimension || h < v.MinImageDimension {
					return nil, fmt.Errorf("image resolution %dx%d is below minimum required resolution of %dx%d pixels", w, h, v.MinImageDimension, v.MinImageDimension)
				}
				widthPtr = &w
				heightPtr = &h
			}
		}
	}

	// Reset reader position for subsequent upload stream
	_, _ = file.Seek(0, io.SeekStart)

	return &FileValidationResult{
		OriginalFilename: originalFilename,
		MIMEType:         detectedMIME,
		Extension:        ext,
		FileSize:         size,
		FileHash:         fileHash,
		Width:            widthPtr,
		Height:           heightPtr,
	}, nil
}

File: backend_api/storage/validator.py
import os
import hashlib
import mimetypes

try:
    from PIL import Image
except ImportError:
    Image = None

class FileValidationResult:
    def __init__(self, original_filename, mime_type, extension, file_size, file_hash, width=None, height=None):
        self.original_filename = original_filename
        self.mime_type = mime_type
        self.extension = extension
        self.file_size = file_size
        self.file_hash = file_hash
        self.width = width
        self.height = height

class FileValidator:
    def __init__(self, max_size_bytes: int = 25 * 1024 * 1024, min_image_dimension: int = 200):
        self.max_size_bytes = max_size_bytes if max_size_bytes > 0 else 10 * 1024 * 1024
        self.min_image_dimension = min_image_dimension if min_image_dimension > 0 else 300

    def validate(self, file, original_filename: str, declared_size: int) -> FileValidationResult:
        ext = os.path.splitext(original_filename)[1].lower()
        blocked_extensions = {".exe", ".sh", ".bat", ".cmd", ".js", ".py", ".php", ".zip", ".tar", ".gz", ".rar", ".7z"}
        if ext in blocked_extensions:
            raise ValueError(f"file type '{ext}' is strictly prohibited for security reasons")

        allowed_extensions = {".pdf", ".jpg", ".jpeg", ".png"}
        if ext not in allowed_extensions:
            raise ValueError(f"unsupported file extension '{ext}'. Only PDF, JPG, JPEG, and PNG files are allowed")

        if declared_size > self.max_size_bytes:
            raise ValueError(f"file size exceeds maximum limit of {self.max_size_bytes // (1024 * 1024)} MB")

        buffer = file.read(512)
        detected_mime = mimetypes.guess_type(original_filename)[0] or "application/octet-stream"
        
        if ext == ".pdf" and "pdf" not in detected_mime and "application/octet-stream" not in detected_mime:
            if len(buffer) >= 4 and buffer[:4] == b"%PDF":
                detected_mime = "application/pdf"
            else:
                raise ValueError("file content does not match valid PDF format")

        allowed_mimes = {"application/pdf", "image/jpeg", "image/jpg", "image/png"}
        if detected_mime not in allowed_mimes and not detected_mime.startswith("image/"):
            raise ValueError(f"invalid file MIME type '{detected_mime}'")

        file.seek(0)
        hasher = hashlib.sha256()
        size = 0
        while chunk := file.read(8192):
            hasher.update(chunk)
            size += len(chunk)

        if size > self.max_size_bytes:
            raise ValueError(f"file size {size} bytes exceeds maximum limit of {self.max_size_bytes // (1024 * 1024)} MB")

        file_hash = hasher.hexdigest()

        width_ptr, height_ptr = None, None
        if detected_mime.startswith("image/"):
            if Image is None:
                raise ValueError("image validation requires Pillow. Install with: pip install Pillow")
            file.seek(0)
            try:
                with Image.open(file) as img:
                    w, h = img.size
                    if w < self.min_image_dimension or h < self.min_image_dimension:
                        raise ValueError(
                            f"image resolution {w}x{h} is below minimum required resolution of "
                            f"{self.min_image_dimension}x{self.min_image_dimension} pixels"
                        )
                    width_ptr, height_ptr = w, h
            except Exception as e:
                if isinstance(e, ValueError):
                    raise e
                raise ValueError(f"unable to read image file: {e}")

        file.seek(0)

        return FileValidationResult(
            original_filename=original_filename,
            mime_type=detected_mime,
            extension=ext,
            file_size=size,
            file_hash=file_hash,
            width=width_ptr,
            height=height_ptr
        )
